File: app/src/main/java/org/mifos/mobilebanking/api/services/GuarantorService.java
package org.mifos.mobilebanking.api.services;

/*
 * Created by saksham on 23/July/2018
 */

import org.mifos.mobilebanking.models.guarantor.GuarantorApplicationPayload;
import org.mifos.mobilebanking.models.guarantor.GuarantorPayload;
import org.mifos.mobilebanking.models.guarantor.GuarantorTemplatePayload;

import java.util.List;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

@SuppressWarnings("PMD.AvoidDuplicateLiterals")
public interface GuarantorService {

    @GET("/loans/{loanId}/guarantors/template")
    Observable<GuarantorTemplatePayload> getGuarantorTemplate(@Path("loanId") long loanId);

    @GET("/loans/{loanId}/guarantors")
    Observable<List<GuarantorPayload>> getGuarantorList(@Path("loanId") long loanId);

    @POST("/loans/{loanId}/guarantors")
    Observable<ResponseBody> createGuarantor(@Path("loanId") long loanId,
                                             @Body GuarantorApplicationPayload payload);

    @PUT("loans/{loanId}/guarantors/{guarantorId}")
    Observable<ResponseBody> updateGuarantor(@Body GuarantorApplicationPayload payload,
                                             @Path("loanId") long loanId,
                                             @Path("guarantorId") long guarantorId);

    @DELETE("/loans/{loanId}/guarantors/{guarantorId}")
    Observable<ResponseBody> deleteGuarantor(@Path("loanId") long loanId,
                                             @Path("guarantorId") long guarantorId);

}

File: app/src/main/java/org/mifos/mobile/api/services/GuarantorService.kt
package org.mifos.mobile.api.services

import okhttp3.ResponseBody
import org.mifos.mobile.models.guarantor.GuarantorApplicationPayload
import org.mifos.mobile.models.guarantor.GuarantorPayload
import org.mifos.mobile.models.guarantor.GuarantorTemplatePayload
import retrofit2.http.*

/*
* Created by saksham on 23/July/2018
*/ interface GuarantorService {
    @GET("loans/{loanId}/guarantors/template")
    suspend fun getGuarantorTemplate(@Path("loanId") loanId: Long?): GuarantorTemplatePayload

    @GET("loans/{loanId}/guarantors")
    suspend fun getGuarantorList(@Path("loanId") loanId: Long?): List<GuarantorPayload>

    @POST("loans/{loanId}/guarantors")
    suspend fun createGuarantor(
        @Path("loanId") loanId: Long?,
        @Body payload: GuarantorApplicationPayload?,
    ): ResponseBody

    @PUT("loans/{loanId}/guarantors/{guarantorId}")
    suspend fun updateGuarantor(
        @Body payload: GuarantorApplicationPayload?,
        @Path("loanId") loanId: Long?,
        @Path("guarantorId") guarantorId: Long?,
    ): ResponseBody

    @DELETE("loans/{loanId}/guarantors/{guarantorId}")
    suspend fun deleteGuarantor(
        @Path("loanId") loanId: Long?,
        @Path("guarantorId") guarantorId: Long?,
    ): ResponseBody
}
